beacon-chain/verification/blob.go

```diff
@@ -0,0 +1,263 @@
+	ErrSidecarParentSeen            = errors.Wrap(ErrBlobInvalid, "parent root has not been seen")
+	ErrSidecarParentValid           = errors.Wrap(ErrBlobInvalid, "parent block is not valid")
+	ErrSidecarParentSlotLower       = errors.Wrap(ErrBlobInvalid, "blob slot <= parent slot")
+	ErrSidecarDescendsFromFinalized = errors.Wrap(ErrBlobInvalid, "blob parent is not descended from the finalized block")
+	ErrSidecarInclusionProven       = errors.Wrap(ErrBlobInvalid, "sidecar inclusion proof verification failure")
+	ErrSidecarBlobCommitmentProven  = errors.Wrap(ErrBlobInvalid, "sidecar kzg commitment proof verification failed")
+	ErrSidecarProposerExpected      = errors.Wrap(ErrBlobInvalid, "sidecar was not proposed by the expected proposer_index")
+)
+
+type BlobVerifier struct {
+	*sharedResources
+	results              *results
+	blob                 blocks.ROBlob
+	parent               state.BeaconState
+	verifyBlobCommitment roblobCommitmentVerifier
+}
+
+type roblobCommitmentVerifier func(blocks.ROBlob) error
+
+// VerifiedROBlob "upgrades" the wrapped ROBlob to a VerifiedROBlob.
+// If any of the verifications ran against the blob failed, or some required verifications
+// were not run, an error will be returned.
+func (bv *BlobVerifier) VerifiedROBlob() (blocks.VerifiedROBlob, error) {
+	if bv.results.allSatisfied() {
+		return blocks.NewVerifiedROBlob(bv.blob), nil
+	}
+	return blocks.VerifiedROBlob{}, bv.results.errors(ErrBlobInvalid)
+}
+
+func (bv *BlobVerifier) recordResult(req Requirement, err *error) {
+	if err == nil || *err == nil {
+		bv.results.record(req, nil)
+		return
+	}
+	bv.results.record(req, *err)
+}
+
+// BlobIndexInBounds represents the follow spec verification:
+// [REJECT] The sidecar's index is consistent with MAX_BLOBS_PER_BLOCK -- i.e. blob_sidecar.index < MAX_BLOBS_PER_BLOCK.
+func (bv *BlobVerifier) BlobIndexInBounds() (err error) {
+	defer bv.recordResult(RequireBlobIndexInBounds, &err)
+	if bv.blob.Index >= fieldparams.MaxBlobsPerBlock {
+		log.WithFields(logging.BlobFields(bv.blob)).Debug("Sidecar index > MAX_BLOBS_PER_BLOCK")
+		return ErrBlobIndexInBounds
+	}
+	return nil
+}
+
+// SlotBelowMaxDisparity represents the spec verification:
+// [IGNORE] The sidecar is not from a future slot (with a MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance)
+// -- i.e. validate that block_header.slot <= current_slot
+func (bv *BlobVerifier) SlotBelowMaxDisparity() (err error) {
+	defer bv.recordResult(RequireSlotBelowMaxDisparity, &err)
+	if bv.clock.CurrentSlot() == bv.blob.Slot() {
+		return nil
+	}
+	// subtract the max clock disparity from the start slot time
+	validAfter := bv.clock.SlotStart(bv.blob.Slot()).Add(-1 * params.BeaconNetworkConfig().MaximumGossipClockDisparity)
+	// If the difference between now and gt is greater than maximum clock disparity, the block is too far in the future.
+	if bv.clock.Now().Before(validAfter) {
+		return ErrSlotBelowMaxDisparity
+	}
+	return nil
+}
+
```

Comment:
TODO: add the gossip spec text to all the other methods.

Reply:
Will this be added in this PR ? Or do you intend to do this in a follow up one

Reply:
https://github.com/prysmaticlabs/prysm/pull/13245/commits/884b5710d81798cf8ffa7d8e7cf1846072ae399a